Dynamic scheduling based on plugins

nateyazdani opened the pull request:
This change set updates the currently available plugins to specify their desired running frequency using the new `statscache.plugins.Frequency` class and to accept a cache timestamp in their `handle()` methods. Note that the `statscache_plugins.releng.Plugin` class did not have any indication of what frequency to use (that I could see), and so I arbitrarily chose 5 minutes for testing.

This pull request restores compatibility with its partner request to the [main statscache engine](https://github.com/fedora-infra/statscache/pull/18).

nateyazdani committed:
Update ``plugin.idx`` to ``plugin.ident``

Update references to the old ``plugin.idx`` property to the new, equivalent
``plugin.ident`` property.

nateyazdani committed:
Make plugin entry-points to classes or collections

There is corresponding change in statscache proper with which this is
co-requisite.

nateyazdani committed:
Include timestamp in handle() methods of plugins

Include the timestamp of the cache in the parameters of the handle() method of
the plugin classes, as specified by the abstract method signature of handle() in
statscache.plugins.BasePlugin. There is an accompanying commit to statscache to
actually provide this argument.

nateyazdani committed:
Switch to statscache.plugins.Frequency

Convert all uses of statscache.utils.{Schedule,Frequency} over to
statscache.plugins.Frequency and add a frequency attribute to all plugins loaded
directly by statscache.

ralphbean commented on statscache_plugins/releng/__init__.py:
```diff
@@ -164,12 +166,13 @@ def initialize(self, session):
         pass
 
     def load_plugins(self, config, model):
+        # TODO: plugins for a plugin? This shoudln't be necessary
```
Ha!  Yeah.  They had some common features to share, but the "plugins on plugins" model may not have been the right fit.  ;)

nateyazdani replied:
Oh whoops, I totally forgot I left that in there! Didn't mean to be so critical :) I'd say there's potential for a "plugin manager" class to fill this sort of purpose

nateyazdani committed:
Use updated statscache.plugins.Frequency class

Update all uses of statscache.plugins.Frequency to reflect the revisions to the
class and its behavior.

nateyazdani committed:
Use interval timedeltas instead of Frequency

In plugin class definitions, an 'interval' attribute is now the means to specify
the desired frequency of cache cycles. Frequency instances are now created all
at once and synchronized by the statscache producer factory. The associated
frequency instance is now provided to the __init__() method of plugin classes
and is made available as a 'frequency' instance attribute by default.

nateyazdani committed:
Update current unit tests

Bring the unit tests up to date with the scheduling improvements to statscache.

ralphbean commented:
:+1:  to merge!
